Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm

```diff
@@ -80,6 +121,77 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
     EGL_DestroySurface(display, handle);
 }
 
+RetainPtr<MTLRasterizationRateMap> newRasterizationRateMap(GCGLDisplay display, IntSize physicalSizeLeft, IntSize physicalSizeRight, IntSize screenSize, std::span<const float> horizontalSamplesLeft, std::span<const float> verticalSamples, std::span<const float> horizontalSamplesRight)
+{
+    EGLDeviceEXT device = EGL_NO_DEVICE_EXT;
+    if (!EGL_QueryDisplayAttribEXT(display, EGL_DEVICE_EXT, reinterpret_cast<EGLAttrib*>(&device)))
+        return nullptr;
+
+    id<MTLDevice> mtlDevice = nil;
+    if (!EGL_QueryDeviceAttribEXT(device, EGL_METAL_DEVICE_ANGLE, reinterpret_cast<EGLAttrib*>(&mtlDevice)))
+        return nullptr;
+
+#if PLATFORM(COCOA) && !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
+    RetainPtr<MTLRasterizationRateMapDescriptor> descriptor = adoptNS([MTLRasterizationRateMapDescriptor new]);
+    id<MTLRasterizationRateMapDescriptorSPI> descriptor_spi = (id<MTLRasterizationRateMapDescriptorSPI>)descriptor.get();
+        descriptor_spi.skipSampleValidationAndApplySampleAtTileGranularity = YES;
+    descriptor_spi.mutability = MTLMutabilityDefault;
+    descriptor_spi.minFactor  = 0.01;
+
+    auto mtlScreenSize = MTLSizeMake(screenSize.width(), screenSize.height(), 0);
+    if (horizontalSamplesLeft.size() > mtlScreenSize.width || horizontalSamplesRight.size() > mtlScreenSize.width || verticalSamples.size() > mtlScreenSize.height)
+        return nullptr;
+
+    RetainPtr<MTLRasterizationRateLayerDescriptor> layerDescriptorLeft = [[MTLRasterizationRateLayerDescriptor alloc] initWithSampleCount:mtlScreenSize];
+
+    memcpy([layerDescriptorLeft horizontalSampleStorage], horizontalSamplesLeft.data(), horizontalSamplesLeft.size_bytes());
```

Comment:
I definitely don't know enough to know that these are safe (and don't obviously see assertions that the sizes of these things are equal)

Reply:
it is a few lines up:
```
    if (horizontalSamplesLeft.size() > mtlScreenSize.width || horizontalSamplesRight.size() > mtlScreenSize.width || verticalSamples.size() > mtlScreenSize.height)
        return nullptr;
```